### PR TITLE
Create DiscoverEC2 User Tasks when Auto Discover fails on EC2 instances

### PR DESCRIPTION
**Pull request by marcoandredinis**
This PR changes the DiscoveryService to start creating and updating
Discover EC2 User Tasks.

So, what are Discover EC2 User Tasks?
When users set up Auto Discover for EC2 Instances, they don't have a
good way of checking for issues on their configured matchers.

We created User Tasks as a way to warn Users that something's wrong.
Each User Task should describe an issue that happened and a way to fix
it.
This has potential to be used to report unexpected events trough the
whole system, which are not errors per se, but something the user should
take action in order to improve the situation.
In this case, we are creating a sub type of those tasks: DiscoverEC2.

From now on, when the DiscoveryService fails to auto-enroll an instance,
it will create a DiscoverEC2 User Task grouping all the failed instances
by the following props:
- integration
- issue type
- account id
- region

A follow up PR will also create notifications so that the user can
actually be notified on those User Tasks and take action.

Demo:
```
$ tctl get user_tasks
kind: user_task
metadata:
  expires:
    nanos: 542005000
    seconds: 1727976384
  name: 52c33845-a413-5b2e-bfa3-1afc5d79236a
  revision: 663ed34b-6c65-4976-9052-af5898b295e1
spec:
  discover_ec2:
    account_id: "123456789012"
    instances:
      i-123:
        discovery_config: dc001
        discovery_group: aws-prod
        instance_id: i-123
        invocation_url: https://eu-west-2.console.aws.amazon.com/systems-manager/run-command/5041ad35-deab-4be7-8bfb-cd64f1cc2a24/i-123
        sync_time:
          nanos: 525000000
          seconds: 1727976024
    region: eu-west-2
  integration: teleportdev
  issue_type: ec2-ssm-script-failure
  state: OPEN
  task_type: discover-ec2
version: v1
---
kind: user_task
metadata:
  expires:
    nanos: 540832000
    seconds: 1727976378
  name: 7da201b0-57d1-503b-856e-de8ca5acd1e1
  revision: 9f803bb5-7ef6-4bd6-a2e5-fbbf8119248a
spec:
  discover_ec2:
    account_id: "123456789012"
    instances:
      i-1234:
        discovery_config: dc001
        discovery_group: aws-prod
        instance_id: i-1234
        sync_time:
          nanos: 379000000
          seconds: 1727976018
    region: eu-west-2
  integration: teleportdev
  issue_type: ec2-ssm-unsupported-os
  state: OPEN
  task_type: discover-ec2
version: v1
---
kind: user_task
metadata:
  expires:
    nanos: 782998000
    seconds: 1727976378
  name: b64397ac-764d-5d9d-af09-30d4775f7a78
  revision: 30ac5a71-a314-43de-a190-ee7db6cfa36a
spec:
  discover_ec2:
    account_id: "123456789012"
    instances:
      i-12345:
        discovery_config: dc001
        discovery_group: aws-prod
        instance_id: i-12345
        sync_time:
          nanos: 631000000
          seconds: 1727976018
    region: eu-west-2
  integration: teleportdev
  issue_type: ec2-ssm-agent-connection-lost
  state: OPEN
  task_type: discover-ec2
version: v1
---
kind: user_task
metadata:
  expires:
    nanos: 291505000
    seconds: 1727976378
  name: dcdca1c5-d489-58bd-8f01-6477ad9756d1
  revision: 176e6527-5cf7-43a2-b1b5-39cdff720f96
spec:
  discover_ec2:
    account_id: "123456789012"
    instances:
      i-123456:
        discovery_config: dc001
        discovery_group: aws-prod
        instance_id: i-123456
        sync_time:
          nanos: 887000000
          seconds: 1727976017
      i-1234567:
        discovery_config: dc001
        discovery_group: aws-prod
        instance_id: i-1234567
        sync_time:
          nanos: 134000000
          seconds: 1727976018
    region: eu-west-2
  integration: teleportdev
  issue_type: ec2-ssm-agent-not-registered
  state: OPEN
  task_type: discover-ec2
version: v1
```

### CODE REVIEW

**Comment by marcoandredinis** [comment]
(i'm working on the flaky test but this should be reviewable already 👍 )

**Comment by marcoandredinis** [comment]
@fheinecke @timothyb89 Can you please take a look when you get a chance?

File: lib/srv/discovery/status.go [hunk 1]
```diff
@@ -293,5 +299,207 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+	}()
+
+	// Fetch the current task because it might have instances discovered by another group of DiscoveryServices.
+	currentUserTask, err := s.AccessPoint.GetUserTask(ctxWithLease, userTaskName)
+	switch {
+	case trace.IsNotFound(err):
+	case err != nil:
+		return trace.Wrap(err)
+	default:
+		for existingIntanceID, existingInstance := range currentUserTask.Spec.DiscoverEc2.Instances {
```
Comment (r0mant) [inline lib/srv/discovery/status.go:439]:
```suggestion
		for existingInstanceID, existingInstance := range currentUserTask.Spec.DiscoverEc2.Instances {
```

File: lib/srv/discovery/status.go [hunk 2]
```diff
@@ -293,5 +299,207 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+			// This ensures that if an instance is removed from AWS, it will eventually disappear from the User Tasks' instance list.
+			// It might also be the case that the DiscoveryConfig was changed and the instance is no longer matched (because of labels/regions or other matchers).
+			instanceIssueExpiration := s.clock.Now().Add(-2 * s.PollInterval)
+			if existingInstance.SyncTime.AsTime().Before(instanceIssueExpiration) {
+				continue
+			}
+
+			// Merge existing backend state into in-memory object.
+			failedInstances[existingIntanceID] = existingInstance
+		}
```
Comment (r0mant) [inline lib/srv/discovery/status.go:457]: I'm having a bit of a hard time parsing this logic. Are we trying to merge instances from this discovery service into existing tasks object? Can it be simplified or refactored somehow to make this more clear? Right now it's a big method that does several things so it's a bit hard to follow.

File: lib/srv/discovery/status.go [hunk 3]
```diff
@@ -293,5 +299,207 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+type awsEC2FailedEnrollmentGroup struct {
+	integration string
+	issueType   string
+	accountID   string
+	region      string
+}
+
+// iterationStarted clears out any in memory issues that were recorded.
+// This is used when starting a new Auto Discover EC2 watcher iteration.
+func (d *awsEC2Tasks) iterationStarted() {
```
Comment (r0mant) [inline lib/srv/discovery/status.go:343]: More of a nit but I think usually methods like this are called something like "reset" or "clear".
```suggestion
func (d *awsEC2Tasks) reset() {
```

File: lib/srv/discovery/status.go [hunk 4]
```diff
@@ -293,5 +299,207 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+
+// awsEC2Tasks contains the Discover EC2 User Tasks that must be reported to the user.
+type awsEC2Tasks struct {
+	mu sync.RWMutex
+	// instancesIssues maps the Discover EC2 User Task grouping parts to a set of instances metadata.
+	instancesIssues map[awsEC2FailedEnrollmentGroup]map[string]*usertasksv1.DiscoverEC2Instance
+	// groupPending is used to register which groups were changed in memory but were not yet sent upstream.
+	// When upserting User Tasks, if the group is not present in groupPending,
+	// then the cluster already has the latest version of this particular group.
+	groupPending map[awsEC2FailedEnrollmentGroup]struct{}
```
Comment (r0mant) [inline lib/srv/discovery/status.go:330]: Nit: groupPending is a pretty confusing name tbh. What is the "group" and why is it "pending"? Can we pick a more descriptive name?

File: lib/srv/discovery/status.go [hunk 5]
```diff
@@ -293,5 +299,207 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+	s.awsEC2Tasks.mu.Lock()
+	defer s.awsEC2Tasks.mu.Unlock()
+	for g := range s.awsEC2Tasks.groupPending {
+		instancesIssueByID := s.awsEC2Tasks.instancesIssues[g]
+		if len(instancesIssueByID) == 0 {
+			continue
+		}
+
+		if err := s.mergeUpsertDiscoverEC2Task(g, instancesIssueByID); err != nil {
+			s.Log.WithError(err).Warning("failed to create discover ec2 user task")
```
Comment (r0mant) [inline lib/srv/discovery/status.go:499]: This is not a very descriptive error message, should we include more info like task we failed to create?

File: lib/srv/discovery/status.go [hunk 6]
```diff
@@ -293,5 +299,207 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+	})
+
+	// Use the deterministic task name as semaphore name.
+	semaphoreName := userTaskName
+	semaphoreExpiration := 5 * time.Second
+
+	// AcquireSemaphoreLock will retry until the semaphore is acquired.
+	// This prevents multiple discovery services to write AWS resources in parallel.
+	// lease must be released to cleanup the resource in auth server.
+	lease, err := services.AcquireSemaphoreLockWithRetry(
```
Comment (r0mant) [inline lib/srv/discovery/status.go:394]: Nit: I usually like to factor out things like this into small helper methods - especially things that are verbose. E.g. this semaphore acquisition is 20 lines of code and it's only tangentially related to the function's logic. This way the caller function is more compact and easier to follow:

```go
func (s *Server) acquireSemaphore(name string) (*services.SemaphoreLock, error) {
    return services.AcuireSemaphoreLockWithRetry(s.ctx, services.SemaphoreLockConfigWithRetry{
        ...
    })
}
```

Then here you will just do:

```go
lease, err := s.acquireSemaphore(userTaskName)
...
```

**Review by r0mant — COMMENTED** [review]
Left some questions and suggestions.

File: lib/srv/discovery/discovery.go [hunk 7]
```diff
@@ -883,6 +888,22 @@ func (s *Server) heartbeatEICEInstance(instances *server.EC2Instances) {
 				discoveryConfig: instances.DiscoveryConfig,
 				integration:     instances.Integration,
 			}, 1)
+
+			s.awsEC2Tasks.addFailedEnrollment(
+				awsEC2FailedEnrollmentGroup{
+					accountID:   instances.AccountID,
+					integration: instances.Integration,
+					issueType:   usertasks.AutoDiscoverEC2IssueEICEFailedToCreateNode,
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply lib/srv/discovery/discovery.go:896]: We still support those if users use `tctl edit discovery_config/<name>`
I'll remove those errors anyway given we are deprecating/removing the feature entirely.

File: lib/srv/discovery/status.go (hunk 4 above)
Comment (marcoandredinis) [reply lib/srv/discovery/status.go:330]: We have 4 fields that we use to group instances into a task: integration, issue type, region and account id.
So, the group is the unique key for this resource.
What do you think about `groupsToSync` or `taskKeysToSync`? 
Changing to the latter, but please let me know if that's clearer?

File: lib/srv/discovery/status.go (hunk 2 above)
Comment (marcoandredinis) [reply lib/srv/discovery/status.go]: Yes, that's correct.
It brings the existing task from the cluster's backend and merges it into the existing in-memory state.
And then updates it.

I've moved the merge part into its own method.
Let me know what you think.

File: lib/srv/discovery/discovery.go [hunk 8]
```diff
@@ -960,6 +965,24 @@ func (s *Server) handleEC2RemoteInstallation(instances *server.EC2Instances) err
 			discoveryConfig: instances.DiscoveryConfig,
 			integration:     instances.Integration,
 		}, len(req.Instances))
+
+		for _, instance := range req.Instances {
+			s.awsEC2Tasks.addFailedEnrollment(
+				awsEC2FailedEnrollmentTaskKey{
+					accountID:   instances.AccountID,
+					integration: instances.Integration,
+					issueType:   usertasks.AutoDiscoverEC2IssueInvocationFailure,
```
Comment (r0mant) [inline lib/srv/discovery/discovery.go:974]: > usertasks.AutoDiscoverEC2IssueInvocationFailure

Nit: Should the issue type include something about SSM? Cause this is SSM installation failure, right? Maybe call it `AutoDiscoverEC2SSMFailure` or something.

File: lib/srv/discovery/status.go [hunk 9]
```diff
@@ -293,5 +300,232 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
 
 	s.updateDiscoveryConfigStatus(result.DiscoveryConfig)
 
+	s.awsEC2Tasks.addFailedEnrollment(
```
Comment (r0mant) [inline lib/srv/discovery/status.go:303]: What's the difference between this invocation and the one in discovery.go above? Just trying to understand what errors will be captured by this function vs the other one cause they seem to be both processing results of SSM runs?

File: lib/srv/discovery/status.go [hunk 10]
```diff
@@ -293,5 +300,232 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+
+// awsEC2Tasks contains the Discover EC2 User Tasks that must be reported to the user.
+type awsEC2Tasks struct {
+	mu sync.RWMutex
+	// instancesIssues maps the Discover EC2 User Task grouping parts to a set of instances metadata.
+	instancesIssues map[awsEC2FailedEnrollmentTaskKey]map[string]*usertasksv1.DiscoverEC2Instance
+	// groupsToSync is used to register which groups were changed in memory but were not yet sent to the cluster.
+	// When upserting User Tasks, if the group is not in groupsToSync,
+	// then the cluster already has the latest version of this particular group.
+	taskKeysToSync map[awsEC2FailedEnrollmentTaskKey]struct{}
```
Comment (r0mant) [inline lib/srv/discovery/status.go:331]: I'd just call it `issuesSyncQueue` or something like that.
```suggestion
	// issuesSyncQueue is used to register which groups were changed in memory but were not yet sent to the cluster.
	// When upserting User Tasks, if the group is not in groupsToSync,
	// then the cluster already has the latest version of this particular group.
	issuesSyncQueue map[awsEC2FailedEnrollmentTaskKey]struct{}
```

File: lib/srv/discovery/status.go [hunk 11]
```diff
@@ -293,5 +300,232 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+}
+
+// awsEC2FailedEnrollmentTaskKey identifies a UserTask group.
+type awsEC2FailedEnrollmentTaskKey struct {
```
Comment (r0mant) [inline lib/srv/discovery/status.go:335]: Nit: Can we shorten the name?
```suggestion
// awsEC2TaskKey identifies a UserTask group.
type awsEC2TaskKey struct {
```

File: lib/srv/discovery/status.go [hunk 12]
```diff
@@ -293,5 +300,232 @@ func (s *Server) ReportEC2SSMInstallationResult(ctx context.Context, result *ser
...
+	if d.taskKeysToSync == nil {
+		d.taskKeysToSync = make(map[awsEC2FailedEnrollmentTaskKey]struct{})
+	}
+	d.taskKeysToSync[g] = struct{}{}
+}
+
+// acquireSemaphoreForUserTask tries to acquire a semaphore lock for this user task.
+// It returns a func which must be called to release the lock.
+// It also returns a context which is tied to the lease and will be canceled if the lease ends.
+func (s *Server) acquireSemaphoreForUserTask(userTaskName string) (func(), context.Context, error) {
```
Comment (r0mant) [inline lib/srv/discovery/status.go:379]:
```suggestion
func (s *Server) acquireSemaphoreForUserTask(userTaskName string) (releaseFn func(), ctx context.Context, err error) {
```

**Review by r0mant — APPROVED** [review]

File: lib/srv/discovery/discovery.go (hunk 8 above)
Comment (marcoandredinis) [reply lib/srv/discovery/discovery.go:974]: Renamed them to `AutoDiscoverEC2IssueSSMInvocationFailure`
I've also renamed other fields to match this pattern.

File: lib/srv/discovery/status.go (hunk 9 above)
Comment (marcoandredinis) [reply lib/srv/discovery/status.go:303]: We have 3 places where we generate tasks:
A) we describe the Instance's SSM Agent state and if there's an error (eg, ssm not installed, ssm not online, os not supported)
B) sending the command (`ssm:SendCommand`) fails (wrong permissions, document params issues, ...)
C) after starting the execution, we check the execution result for each instance

A and C are reported using the `ReportEC2SSMInstallationResult`
B is reported in `discovery.go`
https://github.com/gravitational/teleport/blob/aeefe42e8885355843a7e5803e439a3036abb8d2/lib/srv/discovery/discovery.go#L963

**Review by strideynet — APPROVED** [review]
